### PR TITLE
virt-controller, vmi, Expose multus annotations to vmi

### PR DESCRIPTION
**Pull request by RamLavi**
**What this PR does / why we need it**:
This PR copies the following multus annotations to the vmi:
k8s.v1.cni.cncf.io/networks
k8s.v1.cni.cncf.io/networks-status
The reason we want to expose this is to be able to better map sriov interfaces to their PCI-addresses, as reported on [BZ#2070772](https://bugzilla.redhat.com/show_bug.cgi?id=2070772)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
This PR is relying on another PR https://github.com/kubevirt/kubevirt/pull/8176 (commit 1)

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by RamLavi** [comment]
/hold
I think there is no escape from bumping github.com/k8snetworkplumbingwg/network-attachment-definition-client

**Comment by RamLavi** [comment]
> /hold I think there is no escape from bumping github.com/k8snetworkplumbingwg/network-attachment-definition-client

https://github.com/kubevirt/kubevirt/pull/8093
will remove the hold after it is merged

File: pkg/virt-controller/watch/vmi_test.go
Comment (EdDev) [inline pkg/virt-controller/watch/vmi_test.go:1632]: Ohh my... this is an assertion embedded in the mock expecter.
How about do this:

```golang
vmiInterface.EXPECT().Update(gomock.Any()).Do(func(arg interface{}) {
    vmi = arg.(*virtv1.VirtualMachineInstance)
})
Expect(vmi.Annotations).To(HaveKeyWithValue(networkv1.NetworkStatusAnnot, networkStatusAnnotation))
Expect(vmi.Annotations).To(HaveKeyWithValue(networkv1.NetworkAttachmentAnnot, networkAnnotation))
```

**Review by EdDev — CHANGES_REQUESTED** [review]

File: pkg/virt-controller/watch/vmi.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi.go:770]: I agree, we need to act in case of error. Done

File: pkg/virt-controller/watch/vmi.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi.go:775]: Done

File: pkg/virt-controller/watch/vmi.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi.go:2055]: correct. Done

File: pkg/virt-controller/watch/vmi.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi.go:767]: moved to annotations.go

File: pkg/virt-controller/watch/vmi.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi.go:952]: moved to annotations.go

File: pkg/virt-controller/watch/vmi.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi.go:2067]: > One of the annotations is deprecated, right?

no. both are relevant. 

> I prefer we handle the "merge" here and not later on in the virt-handler. Is that possible?

There is nothing to merge. either the annotation is there, or it's gone.

File: pkg/virt-controller/watch/vmi_test.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi_test.go:1610]: the context is there to increase readability. Initially I added BeforeEach for the pod and vmi preparations but the test does not like it so I left it on the `It` like the all the other tests.
But I think that we still should keep the Context as it makes it easier to understand the test should it fail.

File: pkg/virt-controller/watch/vmi_test.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi_test.go:1612]: I can encapsulate in a function, but IMO the `By` should stay.
`By` adds a much needed visibility to the test output, should the test fail: If the test fails inside a function, the failure will simply point to the test failed, with no indication that it failed in this/that function. 
However, `By` adds `STEP: creating a vmi in Scheduling phase` prints to the test output that helps the developer better triangulate when something went awry

File: pkg/virt-controller/watch/vmi_test.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi_test.go:1632]: Like!

**Comment by RamLavi** [comment]
@EdDev Added a back-off executor to better manage the way the controller is copying the annotations - now it should timeout after 1 second.
Also added unit test to cover migration scenario

File: pkg/virt-controller/watch/vmi.go
Comment (RamLavi) [inline pkg/virt-controller/watch/vmi.go:158]: @EdDev IIUC this backoff should allow a check every 10ms with a mac of 100 steps. Do you feel that we should usea a smaller resolution?

File: pkg/virt-controller/watch/vmi.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi.go]: Done

**Comment by RamLavi** [comment]
retest

**Comment by RamLavi** [comment]
> Hi @RamLavi, as we've talked offline, IMO we only need the `k8s.v1.cni.cncf.io/network-status` annotation in order to solve the SR-IOV PCI addresses mapping problem.

[Done](https://github.com/kubevirt/kubevirt/compare/a5089d49216162a636fd53fd5a859b7641da9238..6a0f9725bc73344cbe8b9f0561bdcc9ddcb50362)

File: pkg/virt-controller/watch/annotations.go
Comment (orelmisan) [inline pkg/virt-controller/watch/annotations.go:34]: Hi @RamLavi, can you please elaborate on the need to use `rateLimitedExecutor`?
Isn't the VMI getting re-queued in case this function fails?

Comment (RamLavi) [reply pkg/virt-controller/watch/annotations.go:40]: good question. it's would give the same result but I am wondering whether we want to "err" the entire controller reconcile over this.. @EdDev what are you thoughts?

File: pkg/virt-controller/watch/vmi.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi.go:542]: this question is related to the former one I think. Let's wait for Eddy say on it

**Comment by RamLavi** [comment]
[reduced](https://github.com/kubevirt/kubevirt/compare/0d37d9bb1f6b0e8abad9aa6ea32c1e028b63cc5d..52a00e1786059f9fb9f013e86bbca486ad6d3f46) the executer timeout to 0.5 to check if this has an effect on the sriov tests failing on the kind lane

**Comment by RamLavi** [comment]
Introduced another PR beneath this one: https://github.com/kubevirt/kubevirt/pull/8126

File: pkg/virt-controller/watch/annotations.go
Comment (orelmisan) [inline pkg/virt-controller/watch/annotations.go:56]: As we've talked offline, another possible solution is to add a returned value RateLimitedExecutor.Exec(), so the caller could distinguish between two scenarios:
1. The `command` function was executed and completed successfully.
2. The `command` function was not executed, because the backoff was not ready.

With this added functionality, the second call to `annotationsCopiedToVmi()` could be dropped.

Comment (RamLavi) [reply pkg/virt-controller/watch/annotations.go:56]: Yes, I agree but I think in this PR we can make due with the current functionality of the rateLimitedExecutor.

File: pkg/virt-controller/watch/vmi_test.go
Comment (orelmisan) [inline pkg/virt-controller/watch/vmi_test.go:1610]:
```suggestion
		Context("The VMI has an SR-IOV interface and the virt-launcher Pod has network-status annotations", func() {
```

File: pkg/virt-controller/watch/vmi_test.go
Comment (orelmisan) [inline pkg/virt-controller/watch/vmi_test.go:1611]:
```suggestion
			DescribeTable("should copy network-status annotation to VirtualMachineInstance and change vmi status to Scheduled", func(oldStatusAnnotation string) {
```

File: pkg/virt-controller/watch/vmi_test.go
Comment (orelmisan) [inline pkg/virt-controller/watch/vmi_test.go:1612]:
```suggestion
				By("creating a vmi in Scheduling phase with an SR-IOV interface")
```

File: pkg/virt-controller/watch/vmi_test.go
Comment (orelmisan) [inline pkg/virt-controller/watch/vmi_test.go:1622]: nit: Why is `new-network-status` surrounded by back-ticks?

File: pkg/virt-controller/watch/vmi_test.go
Comment (orelmisan) [inline pkg/virt-controller/watch/vmi_test.go:1638]: nit:
```suggestion
				Entry("and vmi has no existing network-status annotation", ""),
				Entry("and vmi has an existing network-status annotation", "old-network-status-from-migrated-pod"),
```

File: pkg/virt-controller/watch/vmi_test.go
Comment (orelmisan) [inline pkg/virt-controller/watch/vmi_test.go:1641]:
```suggestion
		Context("VMI does not have an SR-IOV interface and the virt-launcher pod does not have the network-status annotation ", func() {
```

File: pkg/virt-controller/watch/vmi_test.go
Comment (orelmisan) [inline pkg/virt-controller/watch/vmi_test.go:1649]:
```suggestion
				By("attaching a virt-launcher pod without network-status annotation to the vmi")
```

File: pkg/virt-controller/watch/vmi_test.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi_test.go:1610]: fixed but kept the "And .." in the beginning of the Context. I do this to help make the test more readable (should be read as: `when blabla, And The VMI has an SR-IOV interface and the virt-launcher Pod has network-status annotations.`

File: pkg/virt-controller/watch/vmi_test.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi_test.go:1611]: Done

File: pkg/virt-controller/watch/vmi_test.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi_test.go:1612]: changed to `with a`

File: pkg/virt-controller/watch/vmi_test.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi_test.go:1622]: I initially used the full annotation that has " in it, but now that I'm using a dummy content it is not needed. Done

File: pkg/virt-controller/watch/vmi_test.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi_test.go:1638]: Done

File: pkg/virt-controller/watch/vmi_test.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi_test.go:1641]: but it does have an SRIOv interface

Comment (RamLavi) [reply pkg/virt-controller/watch/vmi_test.go:1641]: changed to `And the VMI has an SR-IOV interface and the virt-launcher pod does not have the network-status annotation`

File: pkg/virt-controller/watch/vmi_test.go
Comment (RamLavi) [reply pkg/virt-controller/watch/vmi_test.go:1649]: Done

File: pkg/virt-controller/watch/vmi_test.go
Comment (orelmisan) [reply pkg/virt-controller/watch/vmi_test.go:1641]: My bad, sorry for the confusion.

**Review by EdDev — COMMENTED** [review]
It seems there are several issues that complicates this proposal:
- The mutation of the VMI annotations is done in an area which is suppose to touch non-status fields (although there are some nasty exceptions, I prefer we will not be part of these).
  - I suggest to perform mutation of metadata/spec ations in the `sync` method and in the `updateStatus` to just perform the check if they are present (to implement the wait).
- In the migration scenario, we have a single VMI with two active pods (source and destination) and two virt-handler/s processing the same VMI. In addition we have the migration controller that touches the VMI as well.
  - This is causing races and adds complexity to the solution, something which we would like to avoid.
  
I think we need to go back to the design and re-think this through.

**Comment by RamLavi** [comment]
This PR is no longer considered for merge. Closing

**Pull request closed without merging**
